### PR TITLE
C# Refactor For Webhook Queue

### PR DESCRIPTION
**Pull request by nharper285**
## Summary of the Pull Request

_What is this about?_

## PR Checklist
* [ ] Applies to work item: #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

_How does someone test & validate?_

### CODE REVIEW

File: src/ApiService/ApiService/QueueWebhooks.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+using System.Threading.Tasks;
+using Microsoft.OneFuzz.Service.OneFuzzLib.Orm;
+
+namespace Microsoft.OneFuzz.Service;
+
+public class QueueWebhooks
+{
+    private readonly ILogTracerFactory _loggerFactory;
+
+    public QueueWebhooks(ILogTracerFactory loggerFactory)
```
Comment (stishkin) [inline src/ApiService/ApiService/QueueWebhooks.cs:13]: we do it a bit differently now

see:

https://github.com/microsoft/onefuzz/blob/e32f42cc1bb4a5478449a406d85a9e6125642c15/src/ApiService/ApiService/QueueFileChanges.cs#L20

File: src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs [hunk 2]
```diff
@@ -1,127 +1,127 @@
...
+    const int EXPIRE_DAYS = 7;
+
+    record WebhookMessageQueueObj(
+        Guid WebhookId,
+        Guid EventId
+        );
+
+    private readonly IQueue _queue;
+    private readonly ILogTracer _log;
+    public WebhookMessageLogOperations(IStorage storage, IQueue queue, ILogTracer log) : base(storage)
```
Comment (stishkin) [inline src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs:25]: this is also a good example of how logging works now

Comment (nharper285) [reply src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs:25]: Thanks :) I marked the PR as draft as I'm still working. I needed to create the PR so I could pull down any updates from the rest of the repo.

File: src/ApiService/ApiService/onefuzzlib/Creds.cs [hunk 3]
```diff
@@ -12,11 +13,21 @@ public interface ICreds
...
+
+    public string GetInstanceName();
+
+    public Async.Task<Guid> GetInstanceId();
 }
 
 public class Creds : ICreds
 {
 
+    private IContainers _containers;
```
Comment (chkeita) [inline src/ApiService/ApiService/onefuzzlib/Creds.cs:25]: don't add that dependency. it should be the other way around. Icontainer should depend on Icreds.
Icreds will eventually be a singleton

**Comment by tevoinea** [comment]
There's an empty file `src/ApiService/ApiService/OneFuzzTypes/Primitives.cs`

File: src/ApiService/ApiService/onefuzzlib/Creds.cs (hunk 3 above)
Comment (nharper285) [reply src/ApiService/ApiService/onefuzzlib/Creds.cs:39]: Should I just initialize the container in the GetInstanceId directly?

Comment (nharper285) [reply src/ApiService/ApiService/onefuzzlib/Creds.cs:39]: Moving GetInstanceId to Storage and out of Creds.

File: src/ApiService/ApiService/onefuzzlib/WebhookMessageLogOperations.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    {
+
+        if (message.State == WebhookMessageState.Failed || message.State == WebhookMessageState.Succeeded)
+        {
+            _log.WithTags(
+                new[] {
+                    ("WebhookId", message.WebhookId.ToString()),
+                    ("EventId", message.EventId.ToString()) }
+            ).
+            Error($"webhook message already handled. {message.WebhookId}:{message.EventId}");
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/WebhookMessageLogOperations.cs:86]: Are we missing a return after this `Error` line?

File: src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs [hunk 5]
```diff
@@ -1,125 +1,135 @@
...
+{
+    Async.Task SendEvent(EventMessage eventMessage);
+    Async.Task<Webhook?> GetByWebhookId(Guid webhookId);
+    Async.Task<bool> Send(WebhookMessageLog messageLog);
+}
+
+public class WebhookOperations : Orm<Webhook>, IWebhookOperations
+{
+
+    // Needs to eventually pull from global __version__
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs:19]: We should create an issue to track this

File: src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs [hunk 6]
```diff
@@ -1,125 +1,135 @@
...
+
+        var headers = new Dictionary<string, string> { { "Content-type", "application/json" }, { "User-Agent", USER_AGENT } };
+
+        if (digest != null)
+        {
+            headers["X-Onefuzz-Digest"] = digest;
+        }
+
+        var client = new Request();
+        var response = client.Post(url: webhook.Url, json: data, headers: headers);
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs:84]: Are you missing an `await` here?

```suggestion
        var response = await client.Post(url: webhook.Url, json: data, headers: headers);
```

**Review by tevoinea — APPROVED** [review]
2 small things to check but nothing blocking!

File: src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs (hunk 5 above)
Comment (chkeita) [reply src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs:19]: let's use an app setting for this. i am adding this line in ServiceConfiguration.cs in my current PR
image

We will need to update the build and the deployment to set this up correctly 
CC: @AdamL-Microsoft

Comment (chkeita) [reply src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs:18]: grab th vrsion from Config.OnefuzzVerson

**Review by chkeita — APPROVED** [review]

File: src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs (hunk 5 above)
Comment (nharper285) [reply src/ApiService/ApiService/onefuzzlib/WebhookOperations.cs]: Done!

**Pull request merged**